api/v3/TwingleDonation/Doubleoptinconfirm.php

```diff
@@ -0,0 +1,115 @@
+ *
+ * @param array $params
+ * @return array API result descriptor
+ * @see civicrm_api3_create_success
+ * @see civicrm_api3_create_error
+ */
+function civicrm_api3_twingle_donation_DoubleOptInConfirm($params)
+{
+    // Log call if debugging is enabled within civicrm.settings.php.
+    if (defined('TWINGLE_API_LOGGING') && TWINGLE_API_LOGGING) {
+        CRM_Core_Error::debug_log_message('TwingleDonation.DoubleOptInConfirm: ' . json_encode($params, JSON_PRETTY_PRINT));
+    }
+
+    try {
+        // Get the profile defined for the given form ID, or the default profile
+        // if none matches.
+        $profile = CRM_Twingle_Profile::getProfileForProject($params['project_id']);
+
+        // Get the newsletter groups defined in the profile
+        $newsletter_groups = $profile->getAttribute('newsletter_groups');
+
+        // Extract user email from API call
+        if (!empty($params['user_email'])) {
+            $contacts = civicrm_api3('Email', 'get', array(
+                'seqential' => 1,
+                'email' => $params['user_email'],
+            ));
+
+            // Get pending group memberships for user
+            if (!empty($contacts['values'])) {
+                foreach ($contacts['values'] as $contact) {
+                    $groups = civicrm_api3('GroupContact', 'get', array(
+                        'sequential' => 1,
+                        'contact_id' => $contact['contact_id'],
+                        'status' => "Pending",
+                    ));
+
+                    // Only in newsletter groups: change group membership from pending to added
+                    if (!empty($groups['values'])) {
+                        foreach ($groups['values'] as $group) {
+                            if (in_array($group['group_id'], $newsletter_groups)) {
+                                civicrm_api3('GroupContact', 'create', array(
+                                    'group_id' => $group['group_id'],
+                                    'contact_id' => $contact['contact_id'],
+                                    'status' => "Added",
+                                    $result_values['groups'][] = $group['group_id'],
+                                ));
+                                // Display message if group membership was confirmed correctly
+                                $result_values['double_opt_in'][$group['group_id']] = "Subscription confirmed";
+                            }
+                        }
+                        // Display message if there is no pending group membership
+                    } else {
+                        $result_values['double_opt_in'][] = "Could not confirm subscription: No pending group membership";
+                    }
+                }
+                // Display message if email can't be found
+            } else {
+                $result_values['double_opt_in'][] = "Could not confirm subscription: Email not found";
+            }
+        }
```

Comment:
The contact should be identified using XCM, just as the `submit` API action does. Otherwise, this might alter group memberships for other contacts with the same e-mail address. We should also only target that one single contact identified by XCM with the same profile.

Therefore, all `user_` parameters (and maybe `custom_fields` also) should be added to this API action.

Reply:
Considering that a Double-Opt-In confirmation always refers to an email address, I thought that matching a contact wouldn't be necessary. Especially when the contact has already been matched in the `submit` step. How should we tear DOI confirmations from different contacts with the same email apart, anyway?

Reply:
This is what XCM takes care of: either match a single contact or create one, according to the XCM profile configuration being used. I.e. contacts with the same e-mail address don't matter, since XCM will only return one contact ID. Since `GroupContact` is a thing that belongs to a contact (not an e-mail entity) - it's a crosstab with `group_id` and `contact_id` - we should really make sure we select the correct one and thus identify the previously matched contact again using XCM.